pkg/query/storeset.go

```diff
@@ -167,111 +177,122 @@ func (s *storeRef) close() {
 		externalLabelStores[externalLabelsFromStore(st)]++
 	}
 
 	s.mtx.Lock()
 	defer s.mtx.Unlock()
 
-	// Remove stores that where not updated in this update, because were not in s.peerAddr() this time.
+	// Close stores that where not healthy this time (are not in healthy stores map).
 	for addr, st := range s.stores {
-		if _, ok := stores[addr]; ok {
+		if _, ok := healthyStores[addr]; ok {
 			continue
 		}
 
 		// Peer does not exists anymore.
 		st.close()
+		delete(s.stores, addr)
+
+		level.Info(s.logger).Log("msg", "store unhealthy or does not exists, closed gRPC client and removed from storeset", "address", addr)
 	}
 
-	// Swap old store set with new one, excluding these with duplicated external labels.
-	s.stores = make(map[string]*storeRef, len(stores))
-	for addr, st := range stores {
-		// Check if it has some ext labels specified.
+	// Add stores that are not yet in s.stores.
+	for addr, st := range healthyStores {
+		if _, ok := s.stores[addr]; ok {
+			continue
+		}
+
+		// Check if it has some external labels specified.
 		// No external labels means strictly store gateway or ruler and it is fine to have access to multiple instances of them.
 		//
 		// Sidecar will error out if it will be configured with empty external labels.
-		if len(st.Labels()) == 0 || externalLabelStores[externalLabelsFromStore(st)] == 1 {
-			s.stores[addr] = st
+		if len(st.Labels()) > 0 && externalLabelStores[externalLabelsFromStore(st)] != 1 {
+			st.close()
+			level.Warn(s.logger).Log("msg", "dropping store, external labels are not unique", "address", addr)
 			continue
 		}
 
-		level.Warn(s.logger).Log("msg", "dropping store, external labels are not unique", "address", addr)
-
-		st.close()
+		s.stores[addr] = st
+		level.Info(s.logger).Log("msg", "adding new store to query storeset", "address", addr)
 	}
 
 	s.externalLabelStores = externalLabelStores
 	s.storeNodeConnections.Set(float64(len(s.stores)))
 }
 
+func (s *StoreSet) getHealthyStores(ctx context.Context) map[string]*storeRef {
+	var (
+		unique = make(map[string]struct{})
+
+		healthyStores = make(map[string]*storeRef, len(s.stores))
+		mtx           sync.Mutex
+		g             errgroup.Group
+	)
+
+	// Gather healthy stores map concurrently. Build new store if does not exist already.
+	for _, storeSpec := range s.storeSpecs() {
+		if _, ok := unique[storeSpec.Addr()]; ok {
+			level.Warn(s.logger).Log("msg", "duplicated address in gossip or static store nodes", "address", storeSpec.Addr())
+			continue
+		}
+		unique[storeSpec.Addr()] = struct{}{}
+
+		spec := storeSpec
+		g.Go(func() error {
+			addr := spec.Addr()
+
+			ctx, cancel := context.WithTimeout(ctx, s.gRPCInfoCallTimeout)
+			defer cancel()
+
+			st, ok := s.stores[addr]
+			if ok {
+				// Check existing store. Is it healthy? What are current metadata?
+				labels, minTime, maxTime, err := spec.Metadata(ctx, st.StoreClient)
+				if err != nil {
+					// Peer unhealthy. Do not include in healthy stores.
```

Comment:
I think you shouldn't use `errgroup.Group` here. Using errgroup signals that you have multiple goroutines that should end when one of them errors (which actually doesn't happen in your case). Also `err := g.Wait(); ` will return and print only the first error.
Instead you should use regular `go func()` with explicit waitgroup and log all errors inside the goroutine.